### PR TITLE
fix various issues with describe

### PR DESCRIPTION
**Pull request by rmorshea**
# Summary

Fix various issues with `describe`.

# Notes

These improvements came from #466.

# Todo

Write tests for `describe`.

### CODE REVIEW

File: traitlets/utils/descriptions.py [hunk 1]
```diff
@@ -11,16 +11,16 @@ def describe(article, value, name=None, verbose=False, capital=False):
...
+        one will be infered. Thus, the arguments of ``describe``
         can themselves represent what the resulting string
         will actually look like. If None, then no article
         will be prepended to the result. For non-articled
         description, values that are instances are treated
         definitely, while classes are handled indefinitely.
     value: any
         The value which will be named.
-    name: str or None (default: None)
+    name: str, or None (default: None)
```
Comment (minrk) [inline traitlets/utils/descriptions.py]: typo: no comma

File: traitlets/utils/descriptions.py [hunk 2]
```diff
@@ -164,12 +127,71 @@ def add_article(name, definite=False, capital=False):
...
     if six.PY2 and the_type is types.InstanceType:
         # Old-style class.
         the_type = obj.__class__
     msg = '%r %r' % (obj, the_type)
-    return msg
\ No newline at end of file
+    return msg
+
+
+def _articled(article, value):
```
Comment (minrk) [inline traitlets/utils/descriptions.py]: docstring

File: traitlets/utils/descriptions.py [hunk 3]
```diff
@@ -164,12 +127,71 @@ def add_article(name, definite=False, capital=False):
...
+    if article == "a" or article.endswith(" a"):
+        if value[:1] in "aeiou":
+            article = article[:-1] + "an"
+    elif article == "an" or article.endswith(" an"):
+        if value[:1] not in "aeiou":
+            article = article[:-2] + "a"
+    return "%s %s" % (article, value)
+
+
+def _indefinite(article):
```
Comment (minrk) [inline traitlets/utils/descriptions.py]: docstring

File: traitlets/utils/descriptions.py [hunk 4]
```diff
@@ -164,12 +127,71 @@ def add_article(name, definite=False, capital=False):
...
+            article.endswith(" the") or
+            article.startswith("the ")):
+            return False
+        else:
+            raise ValueError("Expected an indefinite article string that "
+                "ends in 'a' or 'an', or a definite article that starts or "
+                "ends with 'the', not %s" % describe("the", article))
+
+
+def _name(value, verbose):
```
Comment (minrk) [inline traitlets/utils/descriptions.py]: docstring

File: traitlets/utils/descriptions.py [hunk 5]
```diff
@@ -164,12 +127,71 @@ def add_article(name, definite=False, capital=False):
...
+        name = repr(value)
+        verbose = False
+    if verbose:
+        name = _prefix(value) + name
+    if tick_wrap:
+        name = name.join("''")
+    return name
+
+
+def _prefix(value):
```
Comment (minrk) [inline traitlets/utils/descriptions.py]: docstring

File: traitlets/utils/descriptions.py [hunk 6]
```diff
@@ -150,8 +114,7 @@ def add_article(name, definite=False, capital=False):
     if definite:
         result = "the " + name
     else:
-        first_letters = re.compile(r'[\W_]+').sub('', name)
-        if first_letters[:1].lower() in 'aeiou':
+        if name[:1].lower() in 'aeiou':
             result = 'an ' + name
```
Comment (minrk) [inline traitlets/utils/descriptions.py]: should this be using the `_articled` helper?

**Comment by minrk** [comment]
Nice! Looking forward to those tests.

**Comment by Carreau** [comment]
Closing as state for 2 years Labelling accordingly to easily reopen if necessary. Will try to do some cleanup to not have multiple pages of stale PR.

**Pull request closed without merging**